src/main/java/gololang/LazyList.java

```
@@ -0,0 +1,396 @@
+    public String toString() {
+      return "LazyList.EMPTY";
+    }
+  };
+
+  /**
+   * Iterator over a {@code LazyList}.
+   */
+  class LazyListIterator implements Iterator<Object> {
+
+    private LazyList list;
+
+    LazyListIterator(LazyList aLazyList) {
+      this.list = aLazyList;
+    }
+
+    @Override
+    public boolean hasNext() {
+      return !list.isEmpty();
+    }
+
+    @Override
+    public Object next() {
+      Object h = list.head();
+      list = list.tail();
+      return h;
+    }
+
+    @Override
+    public void remove() {
+      throw new UnsupportedOperationException(
+          "remove() is not supported on a lazy list iterator");
+    }
+  }
+
+  private final Object head;
+  private final FunctionReference tail;
+  private LazyList memoTail = null;
+
+  /**
+   * Create a new list from the head and tail values.
+   *
+   * @param head the first value of the list.
+   * @param tail a {@code FunctionReference} that returns a LazyList when invoked.
+   * @return a new {@code LazyList}
+   */
+  public static LazyList cons(Object head, FunctionReference tail) {
+    if (tail == null) {
+      throw new IllegalArgumentException("Use the empty list instead of null as the last element of a LazyList");
+    }
+    return new LazyList(head, tail);
+  }
+
+  private LazyList(Object head, FunctionReference tail) {
+    this.head = head;
+    this.tail = tail;
+  }
+
+  /**
+   * Gets the first element of the list (its head).
+   *
+   * @return an {@code Object}, or {@code null} if the list is empty.
+   */
+  public Object head() {
+    return this.head;
+  }
+
+  /**
+   * Gets the rest of the list (its tail).
+   *
+   * @return a {@code LazyList}, or {@code EMPTY} if the list is empty,
+   * contains only one value, or if the closure failed.
+   */
+  public LazyList tail() {
+    if (memoTail == null) {
+      try {
+        memoTail = (LazyList) (this.tail.invoke());
+      } catch (Throwable e) {
+        memoTail = EMPTY;
```

Why is it ignore an exceptions?  Ignoring exception is the popular cause of bugs, so it should rethrow an exception.

However, what dose it perform when second calling of `tail` method?